### PR TITLE
Add custom header field to each request #8

### PR DESCRIPTION
**Pull request by marcnet80**
https://github.com/iwate/ODataHttpClient/issues/8

### CODE REVIEW

File: src/ODataHttpClient/Models/Request.cs [hunk 1]
```diff
@@ -3,55 +3,77 @@
...
 namespace ODataHttpClient.Models
 {
-    public class Request : IRequest
+	public class Request : IRequest
     {
         internal const string DEFAULT_TYPE_KEY = "odata.type";
         private static Type stringType = typeof(string);
         public HttpMethod Method { get; private set; }
-        public string Uri { get; private set; }
+		public HttpRequestHeaders Headers { get; private set; }
```
Comment (iwate) [inline src/ODataHttpClient/Models/Request.cs]: I think that Headers props is IDictionary<string, string> enough.

Comment (iwate) [reply src/ODataHttpClient/Models/Request.cs]: If this type is HttpRequestMessage, a user need headers helper like src/ODataHttpClient.Tests/Factories/HeadersFactory.cs

I think it is not usiful.

File: src/ODataHttpClient.Tests/Factories/HeadersFactory.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+
+namespace ODataHttpClient.Tests.Factories
+{
+    public static class HeadersFactory
```
Comment (iwate) [inline src/ODataHttpClient.Tests/Factories/HeadersFactory.cs]: I think that this is not needed if Request#Headers's type is IDicitonary<string, string>.

File: src/ODataHttpClient/Models/IResponse.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+using System.Net.Http.Headers;
+
+namespace ODataHttpClient.Models
+{
+    public interface IResponse
+    {
+        HttpRequestHeaders Headers { get; }
```
Comment (iwate) [inline src/ODataHttpClient/Models/IResponse.cs]: I think that HttpResponseHeaders is better for Response#Headers 

https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpresponsemessage?view=netstandard-2.0

Comment (iwate) [reply src/ODataHttpClient/Models/IResponse.cs]: Would you tell me about necessity for Response Interface?
It looks like a useless abstraction for me...

File: src/ODataHttpClient/Models/Request.cs (hunk 1 above)
Comment (iwate) [reply src/ODataHttpClient/Models/Request.cs]: > I think that Headers props is IDictionary<string, string> enough.

OK, I have other idea.

* The prop's type is  HttpRequestHeaders
* But Request factory method's headers prop's type is IDictionary<string, string>

It keep symmetry between Request and Response.

File: src/ODataHttpClient/Models/IResponse.cs (hunk 3 above)
Comment (marcnet80) [reply src/ODataHttpClient/Models/IResponse.cs]: @iwate ok, interface removed.

File: src/ODataHttpClient.Tests/Factories/HeadersFactory.cs (hunk 2 above)
Comment (marcnet80) [reply src/ODataHttpClient.Tests/Factories/HeadersFactory.cs]: @iwate, type remains the same HttpRequestHeaders.

File: src/ODataHttpClient/Models/Request.cs (hunk 1 above)
Comment (marcnet80) [reply src/ODataHttpClient/Models/Request.cs]: @iwate, I choosed the second option, HeadersFactory will create `HttpRequestHeaders` from dictionary. 

I've uncertainty according next changes :
- Remove IRequest#Headers getter interface => it's using for unit tests for headers verification and in client: https://github.com/marcnet80/ODataHttpClient/blob/add-headers/src/ODataHttpClient/ODataClient.cs#L84
I think we should leave this property.

- Remove IRequest#CreateMessage(HttpRequestHeaders headers = null) interface =>
Can I add overloading? Interface `IRequest` will look like that:
<pre>
public interface IRequest
{
   HttpRequestHeaders Headers { get; }
   HttpRequestMessage CreateMessage();
   HttpRequestMessage CreateMessage(HttpRequestHeaders headers);
}
</pre>
@iwate, what do you think?

Comment (iwate) [reply src/ODataHttpClient/Models/Request.cs]: > Remove IRequest#Headers getter interface => it's using for unit tests and in client:

I think you can achieve by using cast without adding Headers to IRequest.
```
Assert.True(((Request)request).Headers.Contains(headerKey));
```

> Remove IRequest#CreateMessage(HttpRequestHeaders headers = null) interface =>
> Can I add overloading? Interface IRequest will look like that:

Umm, No you can not. Sorry.
Request is Builder Class for HttpRequestMessage. I do not accept CreateMessage with any option;

Comment (marcnet80) [reply src/ODataHttpClient/Models/Request.cs]: @iwate, ok, removed.

**Pull request merged**